src/systems/particle_emitter/ParticleEmitter.cc

```diff
@@ -0,0 +1,186 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <ignition/msgs/particle_emitter.pb.h>
+
+#include <string>
+
+#include <ignition/common/Profiler.hh>
+#include <ignition/msgs/Utility.hh>
+#include <ignition/plugin/Register.hh>
+
+#include <ignition/gazebo/components/Name.hh>
+#include <ignition/gazebo/components/ParticleEmitter.hh>
+#include <ignition/gazebo/components/Pose.hh>
+#include <ignition/gazebo/Conversions.hh>
+#include <sdf/Material.hh>
+#include "ParticleEmitter.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+// Private data class.
+class ignition::gazebo::systems::ParticleEmitterPrivate
+{
+  /// \brief The particle emitter.
+  public: ignition::msgs::ParticleEmitter emitter;
+};
+
+//////////////////////////////////////////////////
+ParticleEmitter::ParticleEmitter()
+  : System(), dataPtr(std::make_unique<ParticleEmitterPrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void ParticleEmitter::Configure(const Entity &/*_entity*/,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    EntityComponentManager &_ecm,
+    EventManager &/*_eventMgr*/)
+{
+  // Create a particle emitter entity.
+  auto entity = _ecm.CreateEntity();
+  if (entity == kNullEntity)
+  {
+    ignerr << "Failed to create a particle emitter entity" << std::endl;
+    return;
+  }
+
+  // Name.
+  std::string name = "particle_emitter_entity_" + std::to_string(entity);
+  if (_sdf->HasElement("emitter_name"))
+    name = _sdf->Get<std::string>("emitter_name");
+  this->dataPtr->emitter.set_name(name);
+
+  // Type. The default type is point.
+  this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_POINT);
+  std::string type = _sdf->Get<std::string>("type", "point").first;
+  if (type == "box")
+  {
+    this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_BOX);
+  }
+  else if (type == "cylinder")
+  {
+    this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_CYLINDER);
+  }
+  else if (type == "ellipsoid")
+  {
+    this->dataPtr->emitter.set_type(
+      ignition::msgs::ParticleEmitter_EmitterType_ELLIPSOID);
+  }
+  else if (type != "point")
+  {
+    ignerr << "Unknown emitter type [" << type << "]. Using [point] instead"
+           << std::endl;
+  }
+
+  // Pose.
+  ignition::math::Pose3d pose =
+    _sdf->Get<ignition::math::Pose3d>("pose");
+  ignition::msgs::Set(this->dataPtr->emitter.mutable_pose(), pose);
```

Comment:
It seems to me like the `emitter`'s [pose field](https://github.com/ignitionrobotics/ign-msgs/blob/2494142d90caadc44a632f5241ce071952433f94/proto/ignition/msgs/particle_emitter.proto#L59-L60) is redundant, since we set the pose for this entity as a separate component later in this method: https://github.com/ignitionrobotics/ign-gazebo/blob/10988c99f847f57b461b65372c54baa16a906507/src/systems/particle_emitter/ParticleEmitter.cc#L170

Can we remove the emitter message's pose field from `ign-msgs` and just use the entity's pose component here, or does the emitter message itself also need a pose for something else?

Reply:
I'm not sure if it's needed later or not but I think the msg fields are just designed to be consistent with other entities like [models](https://github.com/ignitionrobotics/ign-msgs/blob/ign-msgs6/proto/ignition/msgs/model.proto) and [links](https://github.com/ignitionrobotics/ign-msgs/blob/ign-msgs6/proto/ignition/msgs/link.proto). We can keep them for consistency or just only add them when they are needed later (which will break ABI so will not make it into Dome). I'm ok with either option

Reply:
Personally, I think we should only include what we need/use in order to remove complexity/confusion. So, if we are sure that we don't need a pose field in the emitter message (it seems like we don't based on the other gazebo particle emitter PRs), then I'd like to take it out. Maybe we should wait until the PRs for parts 3 and 4 are merged first before making this change.

Reply:
just noticed that these field are needed by part 4 (https://github.com/ignitionrobotics/ign-gazebo/pull/615) so that users can send a msg through ign-transport to update the particle emitter properties.

Reply:
Based on what I am seeing from #615, it looks like [`SceneManager::CreateParticleEmitter` and `SceneManager::UpdateParticleEmitter`](https://github.com/ignitionrobotics/ign-gazebo/blob/4f64296d87f3f3b37bf22b763a0d61dde0b78355/src/rendering/SceneManager.cc#L971-L1102) don't do anything with the emitter's pose. I think this is because [the pose is its own component](https://github.com/ignitionrobotics/ign-gazebo/blob/4f64296d87f3f3b37bf22b763a0d61dde0b78355/src/systems/particle_emitter/ParticleEmitter.cc#L266). So, I think that the pose field actually isn't needed for #615, but can someone else confirm this to be sure?

Reply:
made use of `pose` field in pull request https://github.com/ignitionrobotics/ign-gazebo/pull/615. 471f0d1